File: src/discovery.ts
import { existsSync, lstatSync, readdirSync, readFileSync, statSync } from "node:fs";
import type { FileSpec, ManagedFile, ProviderConfig, ProvidersConfig, SkillFile, SkillMetadata } from "./types";
import { expandHomePath, getHomeDir, pathBasename, pathDirname, pathJoin } from "./paths";

const ignoreDir = new Set(["node_modules", ".git", "dist", "build", ".cursor"]);

export function discoverSources(sourceName: string): string[] {
  const cwd = process.cwd();
  const sources: string[] = [];
  walk(cwd, (path, entry) => {
    if (entry.isDirectory() && ignoreDir.has(entry.name)) {
      return "skip";
    }
    if (entry.isFile() && entry.name === sourceName) {
      sources.push(path);
    }
    return "continue";
  });
  return sources;
}

export function discoverAll(
  cfg: ProvidersConfig,
  sourceName: string,
  specSelector: (provider: ProviderConfig) => FileSpec | undefined,
): ManagedFile[] {
  const cwd = process.cwd();
  const allowedDirs = allowedProviderDirs(cfg, specSelector);
  const files: ManagedFile[] = [];

  walk(cwd, (path, entry) => {
    if (entry.isDirectory()) {
      if (ignoreDir.has(entry.name) && !allowedDirs.has(entry.name)) {
        return "skip";
      }
      return "continue";
    }

    const dir = pathDirname(path);
    const filename = entry.name;

    let agent = "";
    if (filename === sourceName) {
      agent = sourceName.toUpperCase().replace(/\.MD$/, "");
    } else {
      for (const [agentName, provider] of Object.entries(cfg.providers)) {
        const spec = specSelector(provider);
        if (!spec) {
          continue;
        }
        if (filename === spec.file) {
          if (!spec.dir) {
            agent = agentName.toUpperCase();
            break;
          }
          if (path.includes(spec.dir)) {
            agent = agentName.toUpperCase();
            break;
          }
        }
      }
      if (!agent) {
        return "continue";
      }
    }

    const stat = lstatSync(path);
    files.push({
      path,
      dir,
      agent,
      file: filename,
      isSymlink: stat.isSymbolicLink(),
      size: stat.size,
    });
    return "continue";
  });

  return files;
}

export function discoverGlobalOnly(cfg: ProvidersConfig): ManagedFile[] {
  const files: ManagedFile[] = [];
  for (const location of globalGuidelinePaths(cfg)) {
    if (!fileExists(location)) {
      continue;
    }
    const info = statSync(location);
    const filename = pathBasename(location);
    const dir = pathDirname(location);
    const agent = inferProviderFromFilename(cfg, filename);
    if (!agent) {
      continue;
    }
    const isSymlink = lstatSync(location).isSymbolicLink();
    files.push({
      path: location,
      dir,
      agent,
      file: filename,
      isSymlink,
      size: info.size,
    });
  }
  return files;
}

export function inferProviderFromFilename(cfg: ProvidersConfig, filename: string): string {
  if (filename === cfg.sources.guidelines) {
    return filename.toUpperCase().replace(/\.MD$/, "");
  }
  for (const [agentName, provider] of Object.entries(cfg.providers)) {
    const spec = provider.guidelines;
    if (spec && filename === spec.file) {
      return agentName.toUpperCase();
    }
  }
  return "";
}

export function fileExists(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

export function isSymlink(path: string): boolean {
  try {
    return lstatSync(path).isSymbolicLink();
  } catch {
    return false;
  }
}

function walk(
  root: string,
  visitor: (path: string, entry: { name: string; isDirectory: () => boolean; isFile: () => boolean }) => "skip" | "continue",
) {
  const entries = readdirSync(root, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = pathJoin(root, entry.name);
    const decision = visitor(fullPath, entry);
    if (decision === "skip") {
      continue;
    }
    if (entry.isDirectory()) {
      walk(fullPath, visitor);
    }
  }
}

function allowedProviderDirs(
  cfg: ProvidersConfig,
  specSelector: (provider: ProviderConfig) => FileSpec | undefined,
): Set<string> {
  const allowed = new Set<string>();
  for (const provider of Object.values(cfg.providers)) {
    const spec = specSelector(provider);
    if (!spec || !spec.dir) {
      continue;
    }
    const parts = spec.dir.split("/");
    if (parts.length > 0) {
      allowed.add(parts[0]);
    }
  }
  return allowed;
}

function globalGuidelinePaths(cfg: ProvidersConfig): string[] {
  return (cfg.globalGuidelines ?? []).map((path) => expandHomePath(path));
}

// Skill discovery functions
export function discoverSkills(): SkillFile[] {
  const skills: SkillFile[] = [];

  // Discover global skills (~/.claude/skills/)
  const globalSkillsDir = pathJoin(getHomeDir(), ".claude", "skills");
  if (existsSync(globalSkillsDir)) {
    const globalSkills = discoverSkillsInDir(globalSkillsDir, "global");
    skills.push(...globalSkills);
  }

  // Discover project skills (.claude/skills/)
  const projectSkillsDir = pathJoin(process.cwd(), ".claude", "skills");
  if (existsSync(projectSkillsDir)) {
    const projectSkills = discoverSkillsInDir(projectSkillsDir, "project");
    skills.push(...projectSkills);
  }

  return skills;
}

export function discoverSourceSkills(sourceDirName: string): string[] {
  const cwd = process.cwd();
  const skillDirs: string[] = [];

  walk(cwd, (path, entry) => {
    if (entry.isDirectory() && ignoreDir.has(entry.name)) {
      return "skip";
    }
    // Skip .claude directory to avoid treating .claude/skills as a source
    if (entry.isDirectory() && entry.name === ".claude") {
      return "skip";
    }
    if (entry.isDirectory() && entry.name === sourceDirName) {
      // Found a source skills directory, scan for skill subdirectories
      const subdirs = readdirSync(path, { withFileTypes: true });
      for (const subdir of subdirs) {
        if (subdir.isDirectory()) {
          const skillFile = pathJoin(path, subdir.name, "SKILL.md");
          if (existsSync(skillFile)) {
            skillDirs.push(pathJoin(path, subdir.name));
          }
        }
      }
      return "skip";
    }
    return "continue";
  });

  return skillDirs;
}

function discoverSkillsInDir(dir: string, location: "global" | "project"): SkillFile[] {
  const skills: SkillFile[] = [];

  try {
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (!entry.isDirectory()) {
        continue;
      }

      const skillName = entry.name;
      const skillPath = pathJoin(dir, skillName);
      const skillFile = pathJoin(skillPath, "SKILL.md");

      if (!existsSync(skillFile)) {
        continue;
      }

      const { metadata, error } = parseSkillMetadata(skillFile);

      skills.push({
        path: skillFile,
        dir: skillPath,
        skillName,
        location,
        metadata,
        error,
      });
    }
  } catch (err) {
    // Directory doesn't exist or can't be read
  }

  return skills;
}

function parseSkillMetadata(skillFile: string): { metadata?: SkillMetadata; error?: string } {
  try {
    const content = readFileSync(skillFile, "utf-8");
    const frontmatterMatch = content.match(/^---\s*\n([\s\S]*?)\n---/);

    if (!frontmatterMatch) {
      return { error: "No frontmatter found" };
    }

    const frontmatter = frontmatterMatch[1];
    const metadata: Partial<SkillMetadata> = {};

    for (const line of frontmatter.split("\n")) {
      const match = line.match(/^(\w+(?:-\w+)*):\s*(.+)$/);
      if (!match) {
        continue;
      }

      const [, key, value] = match;
      const trimmedValue = value.trim();

      if (key === "name") {
        metadata.name = trimmedValue;
      } else if (key === "description") {
        metadata.description = trimmedValue;
      } else if (key === "license") {
        metadata.license = trimmedValue;
      } else if (key === "allowed-tools") {
        metadata.allowedTools = trimmedValue;
      }
    }

    if (!metadata.name || !metadata.description) {
      return { error: "Missing required fields (name, description)" };
    }

    return { metadata: metadata as SkillMetadata };
  } catch (err) {
    return { error: `Failed to parse: ${err}` };
  }
}

File: src/discovery.go
package main

import (
	"os"
	"path/filepath"
	"strings"
)

var ignoreDir = map[string]bool{
	"node_modules": true,
	".git":         true,
	"dist":         true,
	"build":        true,
	".cursor":      true,
}

// discoverSources finds all source files with the given name or pattern
func discoverSources(sourceName string) []string {
	cwd, err := os.Getwd()
	if err != nil {
		return []string{}
	}

	sources := []string{}
	walk(cwd, func(path string, info os.FileInfo) string {
		if info.IsDir() && ignoreDir[info.Name()] {
			return "skip"
		}
		if !info.IsDir() && matchesPattern(info.Name(), sourceName) {
			sources = append(sources, path)
		}
		return "continue"
	})

	return sources
}

// matchesPattern checks if filename matches the given pattern (supports wildcards)
func matchesPattern(filename, pattern string) bool {
	if filename == pattern {
		return true
	}

	if strings.Contains(pattern, "*") {
		matched, _ := filepath.Match(pattern, filename)
		return matched
	}

	return false
}

func discoverAll(cfg *ToolsConfig, sourceName string, specSelector func(ToolConfig) *FileSpec) []ManagedFile {
	cwd, err := os.Getwd()
	if err != nil {
		return []ManagedFile{}
	}

	allowedDirs := allowedToolDirs(cfg, specSelector)
	files := []ManagedFile{}

	walk(cwd, func(path string, info os.FileInfo) string {
		if info.IsDir() {
			if ignoreDir[info.Name()] && !allowedDirs[info.Name()] {
				return "skip"
			}
			return "continue"
		}

		dir := pathDirname(path)
		filename := info.Name()

		agent := ""
		if matchesPattern(filename, sourceName) {
			standardTool := getStandardTool(cfg)
			if standardTool != nil {
				standardSpec := specSelector(*standardTool)
				if standardSpec != nil && standardSpec.Dir != "" && !strings.Contains(path, standardSpec.Dir) {
					return "continue"
				}
			}
			agent = strings.ToUpper(strings.TrimSuffix(filename, ".md"))
		} else {
			for agentName, tool := range cfg.Tools {
				spec := specSelector(tool)
				if spec == nil {
					continue
				}
				if matchesPattern(filename, spec.File) {
					if spec.Dir == "" {
						agent = strings.ToUpper(agentName)
						break
					}
					if strings.Contains(path, spec.Dir) {
						agent = strings.ToUpper(agentName)
						break
					}
				}
			}
			if agent == "" {
				return "continue"
			}
		}

		// Get lstat to check for symlink
		linfo, err := os.Lstat(path)
		isSymlink := false
		if err == nil {
			isSymlink = linfo.Mode()&os.ModeSymlink != 0
		}

		files = append(files, ManagedFile{
			Path:      path,
			Dir:       dir,
			Tool:      agent,
			File:      filename,
			IsSymlink: isSymlink,
			Size:      info.Size(),
		})

		return "continue"
	})

	return files
}

func discoverGlobalOnly(cfg *ToolsConfig) []ManagedFile {
	files := []ManagedFile{}

	for _, location := range globalGuidelinePaths(cfg) {
		if !fileExists(location) {
			continue
		}

		info, err := os.Stat(location)
		if err != nil {
			continue
		}

		filename := pathBasename(location)
		dir := pathDirname(location)
		agent := inferToolFromFilename(cfg, filename)
		if agent == "" {
			continue
		}

		linfo, err := os.Lstat(location)
		isSymlink := false
		if err == nil {
			isSymlink = linfo.Mode()&os.ModeSymlink != 0
		}

		files = append(files, ManagedFile{
			Path:      location,
			Dir:       dir,
			Tool:      agent,
			File:      filename,
			IsSymlink: isSymlink,
			Size:      info.Size(),
		})
	}

	return files
}

func inferToolFromFilename(cfg *ToolsConfig, filename string) string {
	standardTool := getStandardTool(cfg)
	if standardTool != nil && standardTool.Guidelines != nil && filename == standardTool.Guidelines.File {
		return strings.ToUpper(strings.TrimSuffix(filename, ".md"))
	}

	for agentName, tool := range cfg.Tools {
		if tool.Guidelines != nil && filename == tool.Guidelines.File {
			return strings.ToUpper(agentName)
		}
	}

	return ""
}

func getStandardTool(cfg *ToolsConfig) *ToolConfig {
	if cfg.Standard == "" {
		return nil
	}
	tool, exists := cfg.Tools[cfg.Standard]
	if !exists {
		return nil
	}
	return &tool
}

// fileExists checks if a file exists
func fileExists(path string) bool {
	info, err := os.Stat(path)
	if err != nil {
		return false
	}
	return !info.IsDir()
}

// isSymlink checks if a path is a symlink
func isSymlink(path string) bool {
	info, err := os.Lstat(path)
	if err != nil {
		return false
	}
	return info.Mode()&os.ModeSymlink != 0
}

// walk recursively walks a directory tree
func walk(root string, visitor func(string, os.FileInfo) string) {
	entries, err := os.ReadDir(root)
	if err != nil {
		return
	}

	for _, entry := range entries {
		fullPath := pathJoin(root, entry.Name())
		info, err := entry.Info()
		if err != nil {
			continue
		}

		decision := visitor(fullPath, info)
		if decision == "skip" {
			continue
		}

		if entry.IsDir() {
			walk(fullPath, visitor)
		}
	}
}

func allowedToolDirs(cfg *ToolsConfig, specSelector func(ToolConfig) *FileSpec) map[string]bool {
	allowed := make(map[string]bool)

	for _, tool := range cfg.Tools {
		spec := specSelector(tool)
		if spec == nil || spec.Dir == "" {
			continue
		}

		parts := strings.Split(spec.Dir, "/")
		if len(parts) > 0 {
			allowed[parts[0]] = true
		}
	}

	return allowed
}

func globalGuidelinePaths(cfg *ToolsConfig) []string {
	var paths []string

	for _, tool := range cfg.Tools {
		if tool.Guidelines != nil {
			for _, path := range tool.Guidelines.Global {
				paths = append(paths, expandHomePath(path))
			}
		}
	}

	return paths
}
